### PR TITLE
properly continue from promise in then to next then continuation

### PR DESCRIPTION
**Pull request by hayes**

### CODE REVIEW

File: test/native-promises.tap.js [hunk 1]
```diff
@@ -918,6 +950,472 @@ test('instanceof', function diferTest(t) {
   t.end()
 });
 
+test('then chain with promise', function(t) {
+  var listener = addListner();
+
+  listener.nextName = 'accept'
+  var promise = Promise.accept(10);
+
+  listener.nextName = 'internal promise'
```
Comment (hayes) [inline test/native-promises.tap.js]: missing semicolon

File: test/native-promises.tap.js [hunk 2]
```diff
@@ -918,6 +950,472 @@ test('instanceof', function diferTest(t) {
   t.end()
 });
 
+test('then chain with promise', function(t) {
+  var listener = addListner();
+
+  listener.nextName = 'accept'
```
Comment (hayes) [inline test/native-promises.tap.js]: missing semicolon

File: index.js [hunk 3]
```diff
@@ -305,14 +305,26 @@ if (instrumentPromise) {
 
 function ensureAslWrapper(promise) {
   if (!promise.__asl_wrapper) {
-    promise.__asl_wrapper = wrapCallback(function(ctx, fn, result, next) {
+    promise.__asl_wrapper = wrapCallback(function aslWrapper(ctx, fn, result, next) {
+      ensureAslWrapper(next);
```
Comment (taion) [inline index.js]: What's the benefit of moving the async wrapper creation before invoking the "then" continuation? This looks like it makes the tests a bit more complicated to write.

Comment (hayes) [inline index.js]: It should be bellow and only happen if the callback does not return a
promise, but it was not be having as expected, planing on digging into this
a bit more when I have time.  Been trying to squeeze getting this fixed in
between all my other work, wish I had more time to dedicate to this problem

Comment (taion) [inline index.js]: Got it. Do you mind if I merge, then re-write the commit a little bit to add that back, and undo the changes w/`currentName` v. `nextName` in the tests?

Comment (hayes) [inline index.js]: Go for it!
